File: scripts/validate_cheatsheets.py
import re
import os
import sys

INVALID_PLACEHOLDER_REGEX = re.compile(r'(?:\{\{[a-zA-Z_][a-zA-Z0-9_-]*\}\}|\{[a-zA-Z_][a-zA-Z0-9_-]*\}|\[[a-zA-Z_][a-zA-Z0-9_-]*\])')

def validate_content(content: str) -> list[str]:
    errors = []
    lines = content.splitlines()
    if not lines:
        return errors
    
    end_idx = -1
    # Check if there is front matter
    if lines[0].strip() == '---':
        for i in range(1, len(lines)):
            if lines[i].strip() == '---':
                end_idx = i
                break
        if end_idx == -1:
            errors.append("Malformed front matter: missing closing '---'")
            return errors
        
        # Parse front matter lines
        front_matter_lines = lines[1:end_idx]
        
        parsed = {}
        current_key = None
        
        for line in front_matter_lines:
            stripped = line.strip()
            if not stripped or stripped.startswith('#'):
                continue
            
            # Check for list items
            if line.startswith('  - ') or line.startswith('    - '):
                val = line.split('-', 1)[1].strip()
                if current_key:
                    if not isinstance(parsed[current_key], list):
                        if parsed[current_key] is None or parsed[current_key] == '':
                            parsed[current_key] = []
                        else:
                            parsed[current_key] = [parsed[current_key]]
                    parsed[current_key].append(val)
                else:
                    errors.append("YAML Syntax Error: list item without a key")
            elif ':' in line:
                key, val = line.split(':', 1)
                key = key.strip()
                val = val.strip()
                current_key = key
                
                # Check for syntax errors like unclosed flow collection
                if (val.startswith('[') and not val.endswith(']')) or (val.startswith('{') and not val.endswith('}')):
                    errors.append(f"YAML Syntax Error: unclosed flow collection '{val}'")
                    parsed[key] = val
                elif not val:
                    parsed[key] = None
                else:
                    parsed[key] = val
            else:
                errors.append(f"YAML Syntax Error: invalid line '{line}'")
        
        # Validate required fields
        if 'syntax' not in parsed:
            errors.append("Missing 'syntax' field in front matter")
        elif not parsed['syntax']:
            errors.append("'syntax' field cannot be empty")
            
        if 'tags' not in parsed:
            errors.append("Missing 'tags' field in front matter")
        else:
            tags_val = parsed['tags']
            if not isinstance(tags_val, list):
                errors.append("'tags' must be a list")
                
    # Structure and layout checks
    scan_start = end_idx + 1
    last_type = None
    
    for i in range(scan_start, len(lines)):
        line = lines[i].rstrip('\n')
        stripped = line.strip()
        if not stripped:
            continue
            
        if line.startswith('##'):
            last_type = 'heading'
        elif line.startswith('# '):
            last_type = 'description'
        elif line.startswith('#'):
            last_type = 'comment'
        else:
            # It's a command
            if last_type not in ('description', 'command'):
                errors.append(f"Line {i+1}: Command without description: '{stripped}'")
            
            # Check for invalid placeholder styles
            invalid_matches = INVALID_PLACEHOLDER_REGEX.findall(stripped)
            if invalid_matches:
                errors.append(f"Line {i+1}: Use of non-docopt placeholder style: {', '.join(invalid_matches)}. Prefer '<placeholder_name>'.")
                
            last_type = 'command'
            
    return errors

def get_cheatsheets_dir() -> str:
    script_dir = os.path.dirname(os.path.abspath(__file__))
    root_dir = os.path.dirname(script_dir)
    return os.path.join(root_dir, 'sheets')

def main():
    root_dir = get_cheatsheets_dir()
    files_to_check = []
    
    for entry in os.scandir(root_dir):
        if entry.is_file():
            name = entry.name
            if name.startswith('.'):
                continue
            if name in ('validate_cheatsheets.py', 'test_validate_cheatsheets.py', 'README.md', 'LICENSE'):
                continue
            files_to_check.append(entry.path)
            
    total_errors = 0
    total_files = len(files_to_check)
    valid_files = 0
    
    print(f"Scanning {total_files} cheatsheets...")
    for filepath in sorted(files_to_check):
        filename = os.path.basename(filepath)
        try:
            with open(filepath, 'r', encoding='utf-8') as f:
                content = f.read()
            errors = validate_content(content)
            if errors:
                print(f"[FAIL] {filename}:")
                for err in errors:
                    print(f"  - {err}")
                total_errors += len(errors)
            else:
                valid_files += 1
        except Exception as e:
            print(f"[ERROR] Failed to read {filename}: {e}")
            total_errors += 1
            
    print(f"\nSummary: {valid_files}/{total_files} files passed. {total_errors} total errors.")
    if total_errors > 0:
        sys.exit(1)
    else:
        sys.exit(0)

if __name__ == "__main__":
    main()

File: scripts/validate_cheatsheets.ts
import * as fs from 'fs';
import * as path from 'path';

export const INVALID_PLACEHOLDER_REGEX = /(?:\{\{[a-zA-Z_][a-zA-Z0-9_-]*\}\}|\{[a-zA-Z_][a-zA-Z0-9_-]*\}|\[[a-zA-Z_][a-zA-Z0-9_-]*\])/g;

export function validateContent(content: string): string[] {
    const errors: string[] = [];
    const lines = content.split(/\r?\n/);
    if (lines.length === 0 || (lines.length === 1 && lines[0].trim() === '')) {
        return errors;
    }
    
    let endIdx = -1;
    if (lines[0].trim() === '---') {
        for (let i = 1; i < lines.length; i++) {
            if (lines[i].trim() === '---') {
                endIdx = i;
                break;
            }
        }
        if (endIdx === -1) {
            errors.push("Malformed front matter: missing closing '---'");
            return errors;
        }
        
        const frontMatterLines = lines.slice(1, endIdx);
        const parsed: Record<string, any> = {};
        let currentKey: string | null = null;
        
        for (const line of frontMatterLines) {
            const stripped = line.trim();
            if (!stripped || stripped.startsWith('#')) {
                continue;
            }
            
            if (line.startsWith('  - ') || line.startsWith('    - ')) {
                const val = line.substring(line.indexOf('-') + 1).trim();
                if (currentKey) {
                    if (!Array.isArray(parsed[currentKey])) {
                        if (parsed[currentKey] === undefined || parsed[currentKey] === null || parsed[currentKey] === '') {
                            parsed[currentKey] = [];
                        } else {
                            parsed[currentKey] = [parsed[currentKey]];
                        }
                    }
                    parsed[currentKey].push(val);
                } else {
                    errors.push("YAML Syntax Error: list item without a key");
                }
            } else if (line.includes(':')) {
                const parts = line.split(':');
                const key = parts[0].trim();
                const val = parts.slice(1).join(':').trim();
                currentKey = key;
                
                if ((val.startsWith('[') && !val.endsWith(']')) || (val.startsWith('{') && !val.endsWith('}'))) {
                    errors.push(`YAML Syntax Error: unclosed flow collection '${val}'`);
                    parsed[key] = val;
                } else if (!val) {
                    parsed[key] = null;
                } else {
                    parsed[key] = val;
                }
            } else {
                errors.push(`YAML Syntax Error: invalid line '${line}'`);
            }
        }
        
        if (!('syntax' in parsed)) {
            errors.push("Missing 'syntax' field in front matter");
        } else if (!parsed.syntax) {
            errors.push("'syntax' field cannot be empty");
        }
        
        if (!('tags' in parsed)) {
            errors.push("Missing 'tags' field in front matter");
        } else {
            const tagsVal = parsed.tags;
            if (!Array.isArray(tagsVal)) {
                errors.push("'tags' must be a list");
            }
        }
    }
    
    // Structure and layout checks
    const scanStart = endIdx + 1;
    let lastType: 'heading' | 'description' | 'comment' | 'command' | null = null;
    
    for (let i = scanStart; i < lines.length; i++) {
        const line = lines[i];
        const stripped = line.trim();
        if (!stripped) {
            continue;
        }
        
        if (line.startsWith('##')) {
            lastType = 'heading';
        } else if (line.startsWith('# ')) {
            lastType = 'description';
        } else if (line.startsWith('#')) {
            lastType = 'comment';
        } else {
            // It's a command
            if (lastType !== 'description' && lastType !== 'command') {
                errors.push(`Line ${i + 1}: Command without description: '${stripped}'`);
            }
            
            const invalidMatches = stripped.match(INVALID_PLACEHOLDER_REGEX);
            if (invalidMatches) {
                errors.push(`Line ${i + 1}: Use of non-docopt placeholder style: ${invalidMatches.join(', ')}. Prefer '<placeholder_name>'.`);
            }
            
            lastType = 'command';
        }
    }
    
    return errors;
}

export function getCheatsheetsDir(): string {
    const scriptDir = path.dirname(path.resolve(import.meta.path || __filename));
    const rootDir = path.dirname(scriptDir);
    return path.join(rootDir, 'sheets');
}

export function main() {
    const sheetsDir = getCheatsheetsDir();
    if (!fs.existsSync(sheetsDir)) {
        console.error(`Sheets directory does not exist: ${sheetsDir}`);
        process.exit(1);
    }
    
    const files = fs.readdirSync(sheetsDir);
    const filesToCheck = files.filter(name => {
        if (name.startsWith('.')) return false;
        if (['README.md', 'LICENSE'].includes(name)) return false;
        return fs.statSync(path.join(sheetsDir, name)).isFile();
    });
    
    let totalErrors = 0;
    const totalFiles = filesToCheck.length;
    let validFiles = 0;
    
    console.log(`Scanning ${totalFiles} cheatsheets...`);
    for (const filename of filesToCheck.sort()) {
        const filepath = path.join(sheetsDir, filename);
        try {
            const content = fs.readFileSync(filepath, 'utf-8');
            const errors = validateContent(content);
            if (errors.length > 0) {
                console.log(`[FAIL] ${filename}:`);
                for (const err of errors) {
                    console.log(`  - ${err}`);
                }
                totalErrors += errors.length;
            } else {
                validFiles++;
            }
        } catch (e) {
            console.log(`[ERROR] Failed to read ${filename}: ${e}`);
            totalErrors++;
        }
    }
    
    console.log(`\nSummary: ${validFiles}/${totalFiles} files passed. ${totalErrors} total errors.`);
    if (totalErrors > 0) {
        process.exit(1);
    } else {
        process.exit(0);
    }
}

if (import.meta.main) {
    main();
}
